src/libsyntax/parse/parser/expr.rs

```diff
@@ -1093,10 +1093,40 @@ impl<'a> Parser<'a> {
         let mut attrs = outer_attrs;
         attrs.extend(self.parse_inner_attributes()?);
 
-        let blk = self.parse_block_tail(lo, blk_mode)?;
+        let blk = if let Some(interp_blk) = self.parse_interp_user_fn_block()? {
+            interp_blk
+        } else {
+            self.parse_block_tail(lo, blk_mode)?
+        };
         Ok(self.mk_expr(blk.span, ExprKind::Block(blk, opt_label), attrs))
     }
 
+    /// In interpreter mode, parses the placeholder for the user fn body as the user-provided
+    /// block (not from the source).
+    /// Outside of interpreter mode, simply returns `Ok(None)`.
+    fn parse_interp_user_fn_block(&mut self) -> PResult<'a, Option<P<Block>>> {
+        if let Some(ref interp_user_fn) = self.interp_user_fn {
+            let super::InterpUserFn {
+                ref placeholder,
+                ref body,
+            } = *interp_user_fn.clone();
+            if let token::Ident(name, _) = self.token.kind {
+                if &*name.as_str() == placeholder {
+                    // We are in interpreter mode, and have found the block to use as the
+                    // "user fn" body. Insert the given body for the user fn instead of
+                    // parsing one.
+
+                    self.bump();
+                    self.expect(&token::CloseDelim(token::Brace))?;
+
+                    let blk = body.steal();
```

Comment:
Instead of the `Steal` being moved seemingly solely to support this line, given that we have `&mut` access here, can we make it just a `.take()`? We'd need to drop the Rc/Arc around the InterpUserFn, but that seems fine -- I was surprised to see it earlier.

Reply:
I think this is semantically more correct though, and makes it clearer what the code is doing. Also, see my justification in the summary comment for moving `Steal`.

Reply:
Ah, yes -- I think moving Steal is great, it belongs in data structures much more than in librustc.

I would personally prefer that we avoid it here, though, since any use of steal implies some sort of "oh, we're doing weird things here" which isn't the case here at all. Plus, it'll give you the opportunity to add a diagnostic here for when the option has already been `take()`n from by saying something like "use of <placeholder> twice" or so rather than ICE ing from the steal failure.